recipes/sail/all/conanfile.py

```diff
@@ -32,101 +33,105 @@ class SAILConan(ConanFile):
-                  strip_root=True, destination=self._source_subfolder)
-
-    @functools.lru_cache(1)
-    def _configure_cmake(self):
-        except_codecs = []
-
-        if not self.options.with_avif:
-            except_codecs.append("AVIF")
-        if not self.options.with_gif:
-            except_codecs.append("GIF")
-        if not self.options.with_jpeg2000:
-            except_codecs.append("JPEG2000")
-        if not self.options.with_jpeg:
-            except_codecs.append("JPEG")
-        if not self.options.with_png:
-            except_codecs.append("PNG")
-        if not self.options.with_tiff:
-            except_codecs.append("TIFF")
-        if not self.options.with_webp:
-            except_codecs.append("WEBP")
+        get(self, **self.conan_data["sources"][self.version],
+            strip_root=True, destination=self.source_folder)
 
-        cmake = CMake(self)
-        cmake.definitions["SAIL_BUILD_APPS"] = False
-        cmake.definitions["SAIL_BUILD_EXAMPLES"] = False
-        cmake.definitions["SAIL_BUILD_TESTS"] = False
-        cmake.definitions["SAIL_COMBINE_CODECS"] = True
-        cmake.definitions["SAIL_EXCEPT_CODECS"] = ";".join(except_codecs)
-        cmake.definitions["SAIL_INSTALL_PDB"] = False
-        cmake.definitions["SAIL_THREAD_SAFE"] = self.options.thread_safe
-        cmake.configure(build_folder=self._build_subfolder)
+    def generate(self):
+        enable_codecs = []
 
-        return cmake
+        if self.options.with_avif:
+            enable_codecs.append("avif")
+        if self.options.with_gif:
+            enable_codecs.append("gif")
+        if self.options.with_jpeg2000:
+            enable_codecs.append("jpeg2000")
+        if self.options.with_jpeg:
+            enable_codecs.append("jpeg")
+        if self.options.with_png:
+            enable_codecs.append("png")
+        if self.options.with_tiff:
+            enable_codecs.append("tiff")
+        if self.options.with_webp:
+            enable_codecs.append("webp")
+
+        tc = CMakeToolchain(self)
+        tc.variables["SAIL_BUILD_APPS"] = False
+        tc.variables["SAIL_BUILD_EXAMPLES"] = False
+        tc.variables["SAIL_BUILD_TESTS"] = False
+        tc.variables["SAIL_COMBINE_CODECS"] = True
+        tc.variables["SAIL_ENABLE_CODECS"] = ";".join(enable_codecs)
+        tc.variables["SAIL_INSTALL_PDB"] = False
+        tc.variables["SAIL_THREAD_SAFE"] = self.options.thread_safe
+        # TODO: Remove after fixing https://github.com/conan-io/conan/issues/12012
+        if is_msvc(self):
+            tc.cache_variables["CMAKE_TRY_COMPILE_CONFIGURATION"] = str(self.settings.build_type)
+        # TODO: Remove after fixing https://github.com/conan-io/conan-center-index/issues/13159
+        # C3I workaround to force CMake to choose the highest version of
+        # the windows SDK available in the system
+        if is_msvc(self) and not self.conf.get("tools.cmake.cmaketoolchain:system_version"):
```

Comment:
Is it possible to avoid doing this outside of C3I? I know the Qt 6 implements some special validation logic with `        if os.getenv('NOT_ON_C3I', '0') == '0':` to only evaluate in C3I. This allows local devs to set the variable and override this behavior.

Reply:
Environment should be totally avoided on Conan 2.0, so conan config will be the best and safest approach. About Qt, that variable should be replace by a configuration too. As you should know, env vars are fragile, they can be easily replace, also, depending the user/profile they may change too, so it's not safe using them.